src/liballoc/collections/btree/node.rs

```diff
@@ -554,7 +579,8 @@ impl<'a, K, V> NodeRef<marker::Mut<'a>, K, V, marker::Internal> {
         unsafe {
             ptr::write(self.keys_mut().get_unchecked_mut(idx), key);
             ptr::write(self.vals_mut().get_unchecked_mut(idx), val);
-            self.as_internal_mut().edges.get_unchecked_mut(idx + 1).write(edge.node);
+            let internal_node = &mut *self.as_internal_mut();
```

Comment:
Elsewhere you have a comment for why it is safe to create a reference to the entire node. Could you also add one here?

Reply:
I repeated: "SAFETY: we can access the entire node freely and do no need raw pointers, because we have exclusive access." But it really means: "we assume we have exclusive access, and once Miri rats us out, we'll drop that assumption".

More interesting is the one place (in `descend`) where we do need to use raw pointers: "SAFETY: we need to use raw pointers to and inside the node because we may be invoked during iteration, while the caller clings on to outstanding mutable reference to values."

Reply:
> But it really means: "we assume we have exclusive access, and once Miri rats us out, we'll drop that assumption".

:joy: 

Is there some more solid argument we have? Like, there cannot possible be references into this code that have been handed out to users because this is `push` which requires exclusive access to the entire tree (cannot happen during iteration)?

Reply:
We could put that/another comment in the documentation of those functions, saying that they take their `&mut self` seriously - exclusive access all the way. But it's really the other way around. `descend(self)`, and some methods using it, should not pretend to have exclusive access. They should not be `self` or `&mut self`.

Reply:
That's probably nonsense. It can own a handle, the problem is with the tree it refers to. What I'm getting at is probably that there should be an alternative to `marker::Mut`, some `marker::MutNotReallyExclusive` used for iterators.

Reply:
I think this is vastly superior. It allows rolling back many changes here, because `as_internal_mut` (i.e. the possibility to change edges) is in the exclusive access world, far away from the mutable iterator world. Hold on while I decorate the next dish.